provider/file/file_test.go

```diff
@@ -0,0 +1,196 @@
+package file
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+	"testing"
+
+	"github.com/containous/traefik/safe"
+	"github.com/containous/traefik/types"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	frontends = `
+[frontends]
+  [frontends.frontend1]
+  backend = "backend2"
+    [frontends.frontend1.routes.test_1]
+    rule = "Host:test.localhost"
+  [frontends.frontend2]
+  backend = "backend1"
+  passHostHeader = true
+  entrypoints = ["https"] # overrides defaultEntryPoints
+    [frontends.frontend2.routes.test_1]
+    rule = "Host:{subdomain:[a-z]+}.localhost"
+  [frontends.frontend3]
+  entrypoints = ["http", "https"] # overrides defaultEntryPoints
+  backend = "backend2"
+    rule = "Path:/test"
+    `
+
+	backends = `
+[backends]
+  [backends.backend1]
+    [backends.backend1.circuitbreaker]
+      expression = "NetworkErrorRatio() > 0.5"
+    [backends.backend1.servers.server1]
+    url = "http://172.17.0.2:80"
+    weight = 10
+    [backends.backend1.servers.server2]
+    url = "http://172.17.0.3:80"
+    weight = 1
+  [backends.backend2]
+    [backends.backend2.LoadBalancer]
+      method = "drr"
+    [backends.backend2.servers.server1]
+    url = "http://172.17.0.4:80"
+    weight = 1
+    [backends.backend2.servers.server2]
+    url = "http://172.17.0.5:80"
+    weight = 2
+    `
+
+	frontend1 = `
+[frontends]
+  [frontends.frontend1]
+  backend = "backend1"
+    [frontends.frontend1.routes.test_1]
+    rule = "Host:test.localhost"
+    `
+
+	backend1 = `
+  [backends.backend1]
+    [backends.backend1.circuitbreaker]
+      expression = "NetworkErrorRatio() > 0.5"
+    [backends.backend1.servers.server1]
+    url = "http://172.17.0.2:80"
+    weight = 10
+    [backends.backend1.servers.server2]
+    url = "http://172.17.0.3:80"
+    weight = 1
+    `
+)
+
+func TestProvideSingleFile(t *testing.T) {
+	tempDir := createTempDir(t, "testsingle")
+	defer os.RemoveAll(tempDir)
+
+	tempFileName := path.Join(tempDir, "temp1.toml")
+	tempFile := createFile(t, tempFileName)
+	tempFile.WriteString(frontends + backends)
+	tempFile.Close()
+
+	provider := new(Provider)
+	c := make(chan types.ConfigMessage)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	numFrontends := 3
+	numBackends := 2
+
+	go func() {
+		for {
+			data := <-c
+			assert.Equal(t, "file", data.ProviderName)
+			assert.Equal(t, numFrontends, len(data.Configuration.Frontends))
+			assert.Equal(t, numBackends, len(data.Configuration.Backends))
+			wg.Done()
+		}
+
+	}()
+
+	provider.Filename = tempFile.Name()
+	provider.Watch = true
+	provider.Provide(c, safe.NewPool(context.Background()), nil)
+
+	// Wait for initial message to be tested
+	wg.Wait()
```

Comment:
We should define an upper bound after which we stop waiting for the asserting goroutine to complete. Otherwise, our test could lock up if we make a mistake or otherwise fail to send an event down the `ConfigMessage` channel.

The usual way to achieve this is to define a `time.Timer` of, say, 500 ms or so, and have the test `t.Fail` if it fires. If you also replace the wait group by a done channel that's being closed by the asserting goroutine as soon as it finishes, you can have a simple `select` statements over the timeout and done channels.

Similar for other tests below.